resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsEventHandler.scala

```diff
@@ -0,0 +1,212 @@
+            }
+          // TODO (SPARK-24135) - handle more classes of errors
+          case "error" | "failed" | "succeeded" =>
+            removeExecutorFromSpark(schedulerBackend, updatedPod, execId)
+            // If deletion failed on a previous try, we can try again if resync informs us the pod
+            // is still around.
+            // Delete as best attempt - duplicate deletes will throw an exception but the end state
+            // of getting rid of the pod is what matters.
+            if (!isDeleted(updatedPod)) {
+              Utils.tryLogNonFatalError {
+                kubernetesClient
+                  .pods()
+                  .withName(updatedPod.getMetadata.getName)
+                  .delete()
+              }
+            }
+        }
+      }
+    }
+
+    val currentRunningExecutors = runningExecutors.size
+    val currentTotalExpectedExecutors = totalExpectedExecutors.get
+    if (pendingExecutors.isEmpty && currentRunningExecutors < currentTotalExpectedExecutors) {
+      val numExecutorsToAllocate = math.min(
+        currentTotalExpectedExecutors - currentRunningExecutors, podAllocationSize)
+      val newExecutorIds = new TLongArrayList()
+      val podsToAllocate = mutable.Buffer.empty[Pod]
+      for (_ <- 0 until numExecutorsToAllocate) {
+        val newExecutorId = EXECUTOR_ID_COUNTER.incrementAndGet()
+        val executorConf = KubernetesConf.createExecutorConf(
+          conf,
+          newExecutorId.toString,
+          applicationId,
+          driverPod)
+        val executorPod = executorBuilder.buildFromFeatures(executorConf)
+        val podWithAttachedContainer = new PodBuilder(executorPod.pod)
+          .editOrNewSpec()
+          .addToContainers(executorPod.container)
+          .endSpec()
+          .build()
+        newExecutorIds.add(newExecutorId)
+      }
+      kubernetesClient.pods().create(podsToAllocate: _*)
+      pendingExecutors.addAll(newExecutorIds)
+    }
+  }
+
+  def sendUpdatedPodMetadata(updatedPod: Pod): Unit = {
+    eventQueue.add(Seq(updatedPod))
+  }
+
+  def sendUpdatedPodMetadata(updatedPods: Iterable[Pod]): Unit = {
+    eventQueue.add(updatedPods.toSeq)
+  }
+
+  def setTotalExpectedExecutors(newTotal: Int): Unit = totalExpectedExecutors.set(newTotal)
+
+  private def removeExecutorFromSpark(
+      schedulerBackend: KubernetesClusterSchedulerBackend,
+      updatedPod: Pod,
+      execId: Long): Unit = {
+    // Avoid removing twice from Spark's perspective.
+    if (pendingExecutors.contains(execId) || runningExecutors.contains(execId)) {
+      pendingExecutors.remove(execId)
+      runningExecutors.remove(execId)
+      val exitReason = findExitReason(updatedPod, execId)
+      schedulerBackend.doRemoveExecutor(execId.toString, exitReason)
+    }
+  }
+
+  private def findExitReason(pod: Pod, execId: Long): ExecutorExited = {
+    val exitCode = findExitCode(pod)
+    val (exitCausedByApp, exitMessage) = if (isDeleted(pod)) {
```

Comment:
Not sure if this is 100% accurate - the pod may be evicted by the Kubernetes API if the pod misbehaves, so we should introspect if Kubernetes kicked out the pod because the pod itself did something wrong or if the pod was just deleted by a user or by this Spark application.